File: service-rest/src/test/java/com/bnorm/barkeep/ImmutableSetAdapter.java
// Copyright 2016 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep;

import java.io.IOException;
import java.lang.annotation.Annotation;
import java.lang.reflect.Type;
import java.util.Set;

import com.google.common.collect.ImmutableSet;
import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.JsonReader;
import com.squareup.moshi.JsonWriter;
import com.squareup.moshi.Moshi;
import com.squareup.moshi.Types;

public class ImmutableSetAdapter<T> extends JsonAdapter<ImmutableSet<T>> {

  public static final Factory FACTORY = new Factory() {
    @Override
    public JsonAdapter<?> create(Type type, Set<? extends Annotation> annotations, Moshi moshi) {
      Class<?> rawType = Types.getRawType(type);
      if (!annotations.isEmpty()) {
        return null;
      }
      if (rawType == ImmutableSet.class) {
        return newImmutableSetAdapter(type, moshi).nullSafe();
      }
      return null;
    }
  };

  private final JsonAdapter<T> elementAdapter;

  private ImmutableSetAdapter(JsonAdapter<T> elementAdapter) {
    this.elementAdapter = elementAdapter;
  }

  static <T> JsonAdapter<ImmutableSet<T>> newImmutableSetAdapter(Type type, Moshi moshi) {
    Type elementType = Types.collectionElementType(type, ImmutableSet.class);
    JsonAdapter<T> elementAdapter = moshi.adapter(elementType);
    return new ImmutableSetAdapter<>(elementAdapter);
  }

  @Override
  public ImmutableSet<T> fromJson(JsonReader reader) throws IOException {
    ImmutableSet.Builder<T> builder = ImmutableSet.builder();
    reader.beginArray();
    while (reader.hasNext()) {
      builder.add(elementAdapter.fromJson(reader));
    }
    reader.endArray();
    return builder.build();
  }

  @Override
  public void toJson(JsonWriter writer, ImmutableSet<T> value) throws IOException {
    writer.beginArray();
    for (T element : value) {
      elementAdapter.toJson(writer, element);
    }
    writer.endArray();
  }
}

File: service-rest/src/test/kotlin/com/bnorm/barkeep/service/rest/ImmutableSetAdapter.kt
// Copyright 2017 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.service.rest

import com.google.common.collect.ImmutableSet
import com.squareup.moshi.JsonAdapter
import com.squareup.moshi.JsonReader
import com.squareup.moshi.JsonWriter
import com.squareup.moshi.Moshi
import com.squareup.moshi.Types
import java.io.IOException
import java.lang.reflect.Type

class ImmutableSetAdapter<T> private constructor(private val elementAdapter: JsonAdapter<T>) : JsonAdapter<ImmutableSet<T>>() {

  @Throws(IOException::class)
  override fun fromJson(reader: JsonReader): ImmutableSet<T> {
    val builder = ImmutableSet.builder<T>()
    reader.beginArray()
    while (reader.hasNext()) {
      builder.add(elementAdapter.fromJson(reader))
    }
    reader.endArray()
    return builder.build()
  }

  @Throws(IOException::class)
  override fun toJson(writer: JsonWriter, value: ImmutableSet<T>) {
    writer.beginArray()
    for (element in value) {
      elementAdapter.toJson(writer, element)
    }
    writer.endArray()
  }

  companion object {

    val FACTORY: Factory = Factory { type, annotations, moshi ->
      val rawType = Types.getRawType(type)
      if (!annotations.isEmpty()) {
        return@Factory null
      }
      if (rawType == ImmutableSet::class.java) {
        return@Factory newImmutableSetAdapter<Any>(type, moshi).nullSafe()
      }
      null
    }

    internal fun <T> newImmutableSetAdapter(type: Type, moshi: Moshi): JsonAdapter<ImmutableSet<T>> {
      val elementType = Types.collectionElementType(type, ImmutableSet::class.java)
      val elementAdapter = moshi.adapter<T>(elementType)
      return ImmutableSetAdapter(elementAdapter)
    }
  }
}